compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -920,9 +919,55 @@ trait Implicits:
           if normalImports.isEmpty then suggestedImports else normalImports
         end hiddenImplicitsAddendum
 
-        msg(userDefined.getOrElse(
-          em"no implicit argument of type $pt was found${location("for")}"))() ++
-        hiddenImplicitsAddendum
+        val userDefinedParamMessage = paramSymWithMethodTree.flatMap{ (sym, tree) =>
+          userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map{ rawMsg =>
+            val paramSyms = typeVariablesInScope(sym)
+            val paramTypeRefs = paramSyms.map(_.typeRef.typeConstructor)
+
+            val argTypes = tree match {
+              case TypeApply(Select(qual, _), targs) =>
+                val methodOwnerType = qual.tpe
+                val methodOwner = tree.symbol.owner
+                val methodTypeParams = tree.symbol.paramSymss.flatten.filter(_.isType)
+                val methodTypeArgs = targs.map(_.tpe)
+                paramTypeRefs.map(_.asSeenFrom(methodOwnerType, methodOwner).subst(methodTypeParams, methodTypeArgs))
+              case _ => paramTypeRefs
+            }
+
+            err.userDefinedErrorString(
+              rawMsg,
+              paramSyms.map(_.name.unexpandedName.toString),
+              argTypes
+            )
+          }
+        }
+
+        val userDefinedTypeMessage = userDefinedMsg(pt.typeSymbol, defn.ImplicitNotFoundAnnot).map{ rawMsg =>
+          def extractTypeParams(tpe: Type): List[(String, Type)] =
+            tpe match {
+              case ap @ AppliedType(tycon, _) =>
+                val paramsNames = ap.typeSymbol.typeParams.map(_.name.unexpandedName.toString)
+                val paramsValues = ap.widenExpr.dropDependentRefinement.argInfos
+                paramsNames.zip(paramsValues) ++ extractTypeParams(tycon)
+              case RefinedType(parent, _, refinedInfo) => extractTypeParams(parent) ++ extractTypeParams(refinedInfo)
```

Comment:
Are there any test case for refinement types?

Reply:
While trying to test that deeper I realized that there are potentially some more cases but I'm not sure how some of them should be handled. E.g. the piece of code below
```
@annotation.implicitNotFound("There's no Foo[${A}]") trait Foo[A]
implicitly[Foo]
```
* when compiled with scala 2.13 simply returns the error
```
error: trait Foo takes type parameters
```
* when compiled with dotty from latest master, throws an exception inside the compiler and kills the REPL
* when compiled from this branch, prints the user's custom error message but substituting an empty string for the variable
```
There's no Foo[]
```

Should the behaviour in such a case be exactly the same as for scala 2.13? This is not clear for me because if no message is defined by the user, the behaviour is not the same. E.g. for this code
```
trait Bar[A]
implicitly[Bar]
```
the results are as follows:
* scala 2.13:
```
error: trait Bar takes type parameters
```
* dotty's latest master and this PR:
```
no implicit argument of type Bar was found for parameter ev of method implicitly in object DottyPredef
```
Should the same error be reported as in scala 2.13 then or does `implicitly[Foo]` without a type parameter for `Foo` actually make sense in dotty (I assume by treating a type constructor as a type lambda)? If yes then how should the type variable in Foo be substituted while printing the message?

Reply:
@smarter any thoughts on this?

Reply:
`implicitly` takes a type parameter T upper-bounded by Any, so type constructors like `Foo` are not valid arguments for that method (in Dotty we can use an upper-bound of `AnyKind` to allow that sort of things), so Scala 2.13 is correct to complain (and in fact Dotty also complains if I try to write `identity[List](???)`, it's only when the parameter list is implicit that it does something weird). So we should open an issue to fix Dotty's behavior here but this can be done independently from this PR

Reply:
Issue reported, missing tests added.

Reply:
> missing tests added.

Maybe I missed it but I don't see a test that relies on `extractTypeParams(refinedInfo)`, is it possible to construct an example for that?

Reply:
Right, `++ extractTypeParams(refinedInfo)` seems redundant as we take only type parameters into account, not type members (maybe this could be improved in the future but scala 2.13 doesn't handle that either), and you can't add a new type parameter to a type in a refinement nor can you define a new parameterized trait or class inside a refinement, can you?

Anyway while playing with this I discovered there's an issue when one tries to add a missing implicit message to an argument of a method defined inside a refinement so I'll need to investigate into that.

Reply:
@smarter Is it an expected behaviour that a method defined inside a refinement does not have a symbol assigned to it? A comment in Denotations.scala claims that
```
   *  A denotation might refer to `NoSymbol`. This is the case if the denotation
   *  was produced from a disjunction of two denotations with different symbols
   *  and there was no common symbol in a superclass that could substitute for
   *  both symbols.
```
but it's not clear for me if this should be the only case.

E.g. in the code below
```
trait Foo[A]

type Fooable[A] = {
  def foo(implicit @annotation.implicitNotFound("There's no Foo[${A}]") ev: Foo[A]): Any
}

val x: Fooable[Long] = null
x.foo
```
the tree corresponding to `x.foo` has no symbol so the solution using `tree.symbol.paramSymss` to extract the annotation on an argument from the method's definition doesn't work.

When the code above is compiled with scala 2.13, the error message gets resolved properly but the variable is not substituted
```
cmd3.sc:1: There's no Foo[]
val res3 = x.foo
             ^
Compilation Failed
```

Reply:
> Is it an expected behaviour that a method defined inside a refinement does not have a symbol assigned to it?

Yes, the comment in Denotations is incomplete, structural members (members of refinement types) don't have a symbol (a symbol is supposed to have an owner, but a structural member doesn't have a parent which could be used as an owner, and we can't create a new symbol with a dummy owner every time we see a structural member, because two types with the same refinements are equal so it doesn't make sense that they would have members with different symbols)

Reply:
I think the simplest solution is to disallow implicitNotFound annotations on structural members.

Reply:
In terms of compatibility with scala 2.13 then: should we raise an error in such a case or rather simply ignore the annotation and raise a warning?

Reply:
Warning is fine.